docs/IR.md

```diff
@@ -370,7 +372,10 @@ Which is referenced by the Tensor type message:
   }
 ```
 
-The empty list of dimension sizes, [], is a valid tensor shape, denoting a zero-dimension (scalar) value. A zero-dimension tensor is distinct from a tensor of unknown dimensionality, which is indicated by an absent 'shape' property in the Tensor record. When the shape property is absent for an input, a tensor value of any shape may be passed from the caller. When the shape property is absent for an output, the caller should expect that the output value may be of any shape.
+The empty list of dimension sizes, [], is a valid tensor shape, denoting a zero-dimension (scalar) value. A zero-dimension tensor is distinct from a tensor of unknown dimensionality, which is indicated by an absent 'shape' property in the Tensor record. When the shape property is absent for a node input, a tensor value of any shape may be passed from the caller. When the shape property is absent in the type of a value, it indicates that the corresponding runtime value
+may have any shape. However, types with a missing shape are not permitted in all contexts. In particular,
+the inputs and outputs of a model (top-level graphs) are required to *have* a shape, indicating the rank of inputs and outputs,
+even though the exact dimensions need not be specified.
```

Comment:
We probably don't need the info on top level graph inputs/outputs requiring a shape here as it's specified earlier. 

Are there any other known contexts where a type with a missing shape is permitted? If not should we drop the the 'However, types with a missing shape are not permitted in all contexts.' as well? @gramalingam are you aware of other scenarios where a shape is required?

Reply:
these 2 sentences seem redundant:
When the shape property is absent for a node input, a tensor value of any shape may be passed from the caller. When the shape property is absent in the type of a value, it indicates that the corresponding runtime value
may have any shape.

Can these 2 be combined : 
When the shape property is absent in the type of a value (including node input), it indicates that the corresponding runtime value may have any shape.

Reply:
I sort of agree with the idea underlying Scott's point above. To expand on this: this section is meant to explain how to interpret a shape (or a missing-shape) in a type. This is an explanation of the "class TypeProto". It is not meant to be a statement/descriptions of (further) constraints that may be imposed on the use of a TypeProto in particular contexts ... those belong to a description of those contexts.

On the other hand, if you look at the two issues Jacky mentions as motivation, I feel some redundancy may be okay and it may be okay to repeat some information here to avoid some misunderstanding/confusion. 

What if we rephrase part of the above as below
```
This sub-section describes how to interpret a missing-shape or a shape with missing dimensions etc.
However, specific usage contexts may impose further constraints on a type and shape.
For example, the inputs and outputs of a model (top-level graph) are required to *have* a shape, ... ,
as described in section ...
```

Reply:
Just modified. @gramalingam can you check again whether I did the right way?
Thanks!

Reply:
Thanks @jcwchen. Can we add Ashwini's suggestion to the end (those two lines seem to have gotten deleted, without a replacement): namely,
```
When the shape property is absent in the type of a value (including node input),
it indicates that the corresponding runtime value may have any shape.
```

Reply:
Just added. Thanks for the reminder.